python/lsst/daf/butler/tests/_testRepo.py

```diff
@@ -0,0 +1,243 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+__all__ = ["makeTestRepo", "makeTestCollection", "addDatasetType", "expandUniqueId"]
+
+
+import numpy as np
+
+from lsst.daf.butler import Butler, Config, DatasetType
+
+
+def makeTestRepo(root, dataIds, *, config=None, **kwargs):
+    """Create an empty repository with dummy data IDs.
+
+    Parameters
+    ----------
+    root : `str`
+        The location of the root directory for the repository.
+    dataIds : `~collections.abc.Mapping` [`str`, `iterable`]
+        A mapping keyed by the dimensions used in the test. Each value
+        is an iterable of names for that dimension (e.g., detector IDs for
+        `"detector"`). Related dimensions (e.g., instruments and detectors)
+        are linked arbitrarily.
+    config : `lsst.daf.butler.Config`, optional
+        A configuration for the repository (for details, see
+        `lsst.daf.butler.Butler.makeRepo`). If omitted, creates a repository
+        with default dataset and storage types, but optimized for speed.
+    **kwargs
+        Extra arguments to `lsst.daf.butler.Butler.makeRepo`.
+
+    Returns
+    -------
+    butler : `lsst.daf.butler.Butler`
+        A Butler referring to the new repository. This Butler is provided only
+        for additional setup; to keep test cases isolated, it is highly
+        recommended that each test create its own Butler with a
+        unique run/collection. See `makeTestCollection`.
+
+    Notes
+    -----
+    Since the values in ``dataIds`` uniquely determine the repository's
+    data IDs, the fully linked IDs can be recovered by calling
+    `expandUniqueId`, so long as no other code has inserted dimensions into
+    the repository registry.
+    """
+    if not config:
+        config = Config()
+        config[".datastore.cls"] = "lsst.daf.butler.datastores.inMemoryDatastore.InMemoryDatastore"
+        config[".datastore.checksum"] = False  # In case of future changes
+        config[".registry.db"] = "sqlite:///:memory:"
+    Butler.makeRepo(root, config=config, **kwargs)
```

Comment:
It might be worth clarifying in the documentation that this is the desired pattern. While reading the [current docs](https://pipelines.lsst.io/v/d_2020_02_04/py-api/lsst.daf.butler.Butler.html) shows that you can do this, it's not at all obvious what the advantages/disadvantages of any particular config to `Butler()` are.

Reply:
For most people they can't really tell the difference but for outfile it should be the difference between it working and not working at all. A `Config` is a hierarchical dict. A `ButlerConfig` knows about butler configurations and can expand defaults.

Reply:
Even if if the code I originally wrote didn't work, I don't think there's any way I could have determined on my own that this was the fix.

Reply:
My point is that if you have asked for the configuration file to be written somewhere other than root there is an expectation that you would know to use that outfile in your Butler constructor. In your case since you are not in charge of kwargs and do not want to spend time analyzing them, your only safe option is to use the returned config as the argument to Butler.